File: frontend/demo/component/grid/grid-lazy-column-rendering.ts
import 'Frontend/demo/init'; // hidden-source-line

import { html, LitElement } from 'lit';
import { customElement, state } from 'lit/decorators.js';
import '@vaadin/grid';
import { getPeople } from 'Frontend/demo/domain/DataService';
import type Person from 'Frontend/generated/com/vaadin/demo/domain/Person';
import { applyTheme } from 'Frontend/generated/theme';
import type { GridBodyRenderer } from '@vaadin/grid';

@customElement('grid-lazy-column-rendering')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  @state()
  private items: Person[] = [];

  protected override async firstUpdated() {
    const { people } = await getPeople();
    this.items = people;
  }

  protected override render() {
    return html`
      <!-- tag::snippet[] -->
      <vaadin-grid .items="${this.items}" column-rendering="lazy">
        <!-- end::snippet[] -->
        <vaadin-grid-column frozen .renderer="${this.indexColumnRenderer}"></vaadin-grid-column>

        ${[...Array(100).keys()].map((index) => {
          // Generate 100 columns
          return html`
            <vaadin-grid-column
              data-index="${index}"
              .header="${`Col ${index}`}"
              .renderer="${this.columnRenderer}"
            ></vaadin-grid-column>
          `;
        })}
      </vaadin-grid>
    `;
  }

  private columnRenderer: GridBodyRenderer<Person> = (root, column, { index }) => {
    root.textContent = `${column.dataset.index} - ${index}`;
  };

  private indexColumnRenderer: GridBodyRenderer<Person> = (root, _, { index }) => {
    root.textContent = 'Row ' + index;
  };
}

File: src/main/java/com/vaadin/demo/component/grid/GridLazyColumnRendering.java
package com.vaadin.demo.component.grid;

import java.util.List;

import com.vaadin.demo.domain.Person;
import com.vaadin.flow.component.grid.ColumnRendering;
import com.vaadin.flow.component.grid.Grid;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.data.renderer.LitRenderer;
import com.vaadin.flow.router.Route;
import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.demo.domain.DataService;

@Route("grid-lazy-column-rendering")
public class GridLazyColumnRendering extends Div {

    public GridLazyColumnRendering() {

        Grid<Person> grid = new Grid<>(Person.class, false);
        // tag::snippet[]
        grid.setColumnRendering(ColumnRendering.LAZY);
        // end::snippet[]

        grid.addColumn(LitRenderer.of("Row ${index}")).setFrozen(true);

        // Generate 100 columns
        for (int i = 0; i < 100; i++) {
            grid.addColumn(LitRenderer.of(i + " - ${index}"))
                    .setHeader("Col " + i);
        }

        List<Person> people = DataService.getPeople();
        grid.setItems(people);

        add(grid);
    }

    public static class Exporter // hidden-source-line
            extends DemoExporter<GridLazyColumnRendering> { // hidden-source-line
    } // hidden-source-line
}
